modules/percolator/src/main/java/org/elasticsearch/percolator/QueryAnalyzer.java

```diff
@@ -391,42 +393,73 @@ private static Result handleDisjunction(List<Query> disjunctions, int minimumSho
         return new Result(verified, terms);
     }
 
-    static Set<QueryExtraction> selectBestExtraction(Set<QueryExtraction> extractions1, Set<QueryExtraction> extractions2) {
+    static Set<QueryExtraction> selectBestExtraction(Map<String, Float> boostFields, Set<QueryExtraction> extractions1,
+                                                     Set<QueryExtraction> extractions2) {
         assert extractions1 != null || extractions2 != null;
         if (extractions1 == null) {
             return extractions2;
         } else if (extractions2 == null) {
             return extractions1;
         } else {
+            Set<QueryExtraction> filtered1;
+            Set<QueryExtraction> filtered2;
+            if (boostFields.isEmpty() == false) {
+                Predicate<QueryExtraction> predicate = extraction -> {
+                    String fieldName = extraction.term != null ? extraction.term.field() : extraction.range.fieldName;
+                    float boost = boostFields.getOrDefault(fieldName, 1F);
+                    return boost != 0F;
+                };
+                filtered1 = extractions1.stream().filter(predicate).collect(toSet());
+                if (filtered1.isEmpty()) {
+                    return extractions2;
+                }
+                filtered2 = extractions2.stream().filter(predicate).collect(toSet());
+                if (filtered2.isEmpty()) {
+                    return extractions1;
+                }
+
+                float extraction1HighestBoost = highestBoost(filtered1, boostFields);
+                float extraction2HighestBoost = highestBoost(filtered2, boostFields);
```

Comment:
I'm wondering whether we should compare the highest, or lower boosts?
For instance imagine the query is `(a OR b) AND c`. `a`, `b` and `c` have boosts of `2`, `0.5` and `1` respectively. I'd much rather pick `c` as an extracted query than `a OR b` since the cost of `b` might indicate it is a low-cardinality field, ie. each term has many values?